genet/core.py

```diff
@@ -2200,54 +2202,38 @@ def add_elevation_to_nodes(self, elevation_tif_file_path, null_value: float):
             elevation_dict[node_id] = {'z': z}
 
-        self.apply_attributes_to_nodes(elevation_dict)
+        if run_validation is True:
+            report = elevation.validation_report_for_node_elevation(elevation_dict)
+            logging.info(report['summary'])
+
+        return elevation_dict
 
-    def validation_report_for_node_elevation(self, low_limit=-50, mont_blanc_height=4809):
+    def get_link_slope_dictionary(self, elevation_dict):
         """
-        Generates a validation report for the elevation data added to the network nodes.
-        :param low_limit: set at -50 by default, can optionally set a different value
-        :param mont_blanc_height: defaults to 4809m is the height of Mont Blank, can optionally set a different value
-        :return: dict, with 2 data subsets - summary statistics, and extreme values lists
+        Takes an elevation raster file in .tif format, finds the elevation of the end nodes for each link, calculates
+        link slope and adds returns a dictionary of link IDs and their slopes; can then use
+        self.apply_attributes_to_links() function to add slope as a link attribute to the network.
+        :param elevation_dict: contains node_id as key and elevation in meters as value
+        :return: dict in format {link_id : {'slope': slope}}
         """
-        graph = self.subgraph_on_link_conditions(conditions={'modes': all}, mixed_dtypes=True)
-        gdfs = geojson.generate_geodataframes(graph)
-        nodes = gdfs['nodes']
-
-        if 'z' in list(nodes.columns):
-            elevation_list = nodes['z'].to_list()
-            min_value = np.min(elevation_list)
-            max_value = np.max(elevation_list)
-            mean = np.mean(elevation_list)
-            median = np.median(elevation_list)
-
-            nodes_dictionary = dict(zip(nodes['id'], nodes['z']))
-            too_high = {}
-            too_low = {}
+        slope_dict = {}
 
-            for node_id, node_elev in nodes_dictionary.items():
-                if node_elev < low_limit:
-                    too_low[node_id] = node_elev
-                elif node_elev > mont_blanc_height:
-                    too_high[node_id] = node_elev
+        for link in self.links():
+            link_id = link[0]
+            node_1 = self.link(link_id)['from']
+            node_2 = self.link(link_id)['to']
+            z_1 = elevation_dict[node_1]['z']
+            z_2 = elevation_dict[node_2]['z']
 
-            report = {
-                'summary': {'total_nodes': len(elevation_list),
-                            'min_value': int(min_value),
-                            'max_value': int(max_value),
-                            'mean': int(mean),
-                            'median': int(median),
-                            'extremely_high_values_count': len(too_high),
-                            'extremely_low_values_count': len(too_low)},
+            # TO-DO: calculate crow-fly distance between the 2 nodes
```

Comment:
Yes we are using routed distance now to make it simpler, because calculating crow-fly distance requires us to import another module to get the formula... @KasiaKoz said it would be quite bulky, so we decided to use this as proxy for now